pkg/synthetictests/duplicated_events.go

```diff
@@ -64,18 +64,62 @@ var knownEventProblems = []struct {
 	},
 }
 
+type duplicateEventsEvaluator struct {
+	allowedRepeatedEventPatterns []*regexp.Regexp
+	allowedRepeatedEventFns      []isRepeatedEventOKFunc
+
+	// knownRepeatedEventsBugs are duplicates that are considered bugs and should flake, but not  fail a test
+	knownRepeatedEventsBugs []knownProblem
+}
+
+type knownProblem struct {
+	Regexp *regexp.Regexp
+	BZ     string
+}
+
+func testDuplicatedEventForUpgrade(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	allowedPatterns := []*regexp.Regexp{}
+	allowedPatterns = append(allowedPatterns, allowedRepeatedEventPatterns...)
+	allowedPatterns = append(allowedPatterns, allowedUpgradeRepeatedEventPatterns...)
+
+	evaluator := duplicateEventsEvaluator{
+		allowedRepeatedEventPatterns: allowedPatterns,
+		allowedRepeatedEventFns:      nil,
+		knownRepeatedEventsBugs:      knownEventsBugs,
+	}
+
+	return evaluator.testDuplicatedEvents(events)
+}
+
+func testDuplicatedEventForStableSystem(events monitorapi.Intervals) []*ginkgo.JUnitTestCase {
+	evaluator := duplicateEventsEvaluator{
+		allowedRepeatedEventPatterns: allowedRepeatedEventPatterns,
+		allowedRepeatedEventFns:      nil,
+		knownRepeatedEventsBugs:      knownEventsBugs,
+	}
+
+	return evaluator.testDuplicatedEvents(events)
+}
+
+// isRepeatedEventOKFunc takes a monitorEvent as input and returns true if the repeated event is OK.
+// this commonly happens for known bugs and for cases where events are repeated intentionally by tests.
+// the string is the message to display for the failure.
+type isRepeatedEventOKFunc func(monitorEvent monitorapi.EventInterval) (bool, string)
```

Comment:
nit: thoughts about:

```go
// checkRepeatedEventFunc takes a monitorEvent as input and returns an error if the repeated event is not ok.  Returns nil for known bugs, events repeated intentionally by tests, and other acceptable events.
type checkRepeatedEventFunc func(monitorEvent monitorapi.EventInterval) error
```

So you don't have to explain a `bool, string` return combo?